lib/rules/no-unstable-nested-components.js

```diff
@@ -0,0 +1,434 @@
+/**
+ * @fileoverview Prevent creating unstable components inside components
+ * @author Ari Perkkiö
+ */
+
+'use strict';
+
+const Components = require('../util/Components');
+const docsUrl = require('../util/docsUrl');
+
+// ------------------------------------------------------------------------------
+// Constants
+// ------------------------------------------------------------------------------
+
+const ERROR_MESSAGE_WITHOUT_NAME = 'Declare this component outside parent component or memoize it.';
+const COMPONENT_AS_PROPS_INFO = ' If you want to allow component creation in props, set allowAsProps option to true.';
+const RENDER_REGEXP = /^render/;
+const HOOK_REGEXP = /^use[A-Z0-9].*$/;
```

Comment:
it'd be ideal to avoid hardcoding this, to avoid deviating from eslint-plugin-react-with-hooks. is there any way we could share hook detection code with that plugin?

Reply:
At the moment, no. Currently they are not exporting [the hook detection methods](https://github.com/facebook/react/blob/6b28eb6175e5251b31507749b79c55c0e389b6ef/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L12-L19) - I just decided to copy-paste the implementation. If they did export this, we could rely on that via dependencies/peerDependencies.

It's a shame hook related rules live in a separate package.